core/src/main/scala/org/http4s/util/encoding/EncodedString.scala

```diff
@@ -0,0 +1,50 @@
+package org.http4s.util.encoding
+
+import org.http4s.util.encoding.tags._
+import org.http4s.parser.{Rfc3986CharPredicate => CP}
+import org.http4s.util.encoding.UriCodingUtils.{percentEncode, percentDecode}
+import org.parboiled2.CharPredicate
+
+
+sealed trait EncodedString[E] {
```

Comment:
Didn't run any benchmarks, but I think they can probably be value classes in most cases.